### PR TITLE
Authorization/JAXB-API for Jenkins #55, #57

### PR DESCRIPTION
**Pull request by gnl42**
Added "crumbissuer" authentication. Do we still need the old authentication code?

Fixed missing JAXB-API complaint 
(https://www.vogella.com/blog/eclipse-rcp-java-11-jaxb/)

### CODE REVIEW

File: mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java
Comment (wimjongman) [inline mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:127]: You can delete this.

File: mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java
Comment (wimjongman) [inline mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:170]: Please drop.

File: mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java
Comment (wimjongman) [inline mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:126]: You can drop it.

File: mylyn.builds/org.eclipse.mylyn.hudson.core/META-INF/MANIFEST.MF
Comment (wimjongman) [inline mylyn.builds/org.eclipse.mylyn.hudson.core/META-INF/MANIFEST.MF:13]: Can't you use import package instead of require bundle? Also, are all of these required?

File: mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java
Comment (wimjongman) [inline mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java:330]: Put in JSESSIONID in a constant?

**Review by wimjongman — CHANGES_REQUESTED** [review]
Thanks, George! 

This looks good. I found a couple of nits.

File: mylyn.builds/org.eclipse.mylyn.hudson.core/META-INF/MANIFEST.MF
Comment (ruspl-afed) [inline mylyn.builds/org.eclipse.mylyn.hudson.core/META-INF/MANIFEST.MF:7]: Please keep alphabetic order for dependencies

File: org.eclipse.mylyn-target/mylyn-e4.26.target
Comment (ruspl-afed) [inline org.eclipse.mylyn-target/mylyn-e4.26.target:85]: Why do we need this one?

File: org.eclipse.mylyn-target/mylyn-e4.26.target
Comment (ruspl-afed) [inline org.eclipse.mylyn-target/mylyn-e4.26.target:80]: we have this in the target already from `https://download.eclipse.org/tools/orbit/downloads/drops/R20221123021534`

File: org.eclipse.mylyn-target/mylyn-e4.26.target
Comment (ruspl-afed) [inline org.eclipse.mylyn-target/mylyn-e4.26.target:75]: Would be better to use one from `https://download.eclipse.org/tools/orbit/downloads/drops/R20221123021534`

**Review by ruspl-afed — CHANGES_REQUESTED** [review]

**Comment by gnl42** [comment]
Hi Wim, 
  I made the changes and had to rework the code a bit to fix some issues I came across after some more testing.

The build framework is weird, I don't understand what the intent was in its design.
CommonHttpClient has an "authenticated" flag that is good for a session, yet has a ThreadLocal variable "context" that is used to store the token. Side effect of this is that each thread had to authenticate rather than just doing it once for the session.

One issue I'm unable to solve is how use the JSESSION id to authenticate when running a build using ones password.
It works fine when using a API token, but according to all the documentation I've come across it should be doable in both cases???

I've been testing it against my local Jenkins instance, and against https://ci.eclipse.org/mylyn/job/github-mylyn/.
If someone who is able to login to ci.eclipse.org could test the code I would feel more comfortable.

Also, I haven't seen any sign that the crumb is needed. Could be my jenkins hasn't been set up fully.

File: org.eclipse.mylyn-target/mylyn-e4.26.target
Comment (gnl42) [reply org.eclipse.mylyn-target/mylyn-e4.26.target:84]: I have no idea how it's all wired together. Maybe https://www.vogella.com/blog/eclipse-rcp-java-11-jaxb/ will make more sense to you?

Comment (ruspl-afed) [reply org.eclipse.mylyn-target/mylyn-e4.26.target:84]: Yes, true, seems to be required to activate plain java services expected by JAXB impl with OSGi. We can get this bundle from https://download.eclipse.org/tools/orbit/downloads/drops/R20221123021534/ as well.

**Comment by BeckerFrank** [comment]
> Also, I haven't seen any sign that the crumb is needed. Could be my jenkins hasn't been set up fully.

I have finally remember the password so that I could reactivate the Hudson and Jenkins test instances of the http://mylyn.org domain. Maybe that helps.

File: org.eclipse.mylyn-target/mylyn-e4.26.target
Comment (gnl42) [reply org.eclipse.mylyn-target/mylyn-e4.26.target]: Hmm, when I moved it to orbit, the build breaks with:
`Error:  Cannot resolve project dependencies:
Error:    Software being installed: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier
Error:    Missing requirement: org.eclipse.mylyn.commons.ui 3.26.0.qualifier requires 'osgi.bundle; org.apache.commons.lang [2.6.0,3.0.0)' but it could not be found
Error:    Cannot satisfy dependency: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier depends on: org.eclipse.equinox.p2.iu; org.eclipse.mylyn.commons.ui 0.0.0`

Comment (gnl42) [reply org.eclipse.mylyn-target/mylyn-e4.26.target:74]: Never mind. It's only the source that's available from orbit. Sticking to the maven version for now

Comment (gnl42) [reply org.eclipse.mylyn-target/mylyn-e4.26.target:74]: Orbit searches are weird. I did find the right one, but it has the same problem.
I also tried the 2.3.0 version, but it has similar, but different, problems

File: org.eclipse.mylyn-target/mylyn-e4.26.target
Comment (gnl42) [reply org.eclipse.mylyn-target/mylyn-e4.26.target:84]: Hmm, when I moved it to orbit, the build breaks with:
Error:  Cannot resolve project dependencies: Error:    Software being installed: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier Error:    Missing requirement: org.eclipse.mylyn.commons.ui 3.26.0.qualifier requires 'osgi.bundle; org.apache.commons.lang [2.6.0,3.0.0)' but it could not be found Error:    Cannot satisfy dependency: org.eclipse.mylyn.commons.feature.group 3.26.0.qualifier depends on: org.eclipse.equinox.p2.iu; org.eclipse.mylyn.commons.ui 0.0.0

I also tried the 2.3.0 version, but it has similar, but different, problems

Comment (gnl42) [reply org.eclipse.mylyn-target/mylyn-e4.26.target:84]: Tried to change it to commons.lang3 but it's not available from orbit(?)

File: mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java
Comment (gnl42) [reply mylyn.builds/org.eclipse.mylyn.hudson.core/src/org/eclipse/mylyn/internal/hudson/core/client/HudsonOperation.java]: Alas, the Hudson/Jenkins servers used for the unit tests doesn't support the new way, so I put them back

**Comment by gnl42** [comment]
> > Also, I haven't seen any sign that the crumb is needed. Could be my jenkins hasn't been set up fully.
> 
> I have finally remember the password so that I could reactivate the Hudson and Jenkins test instances of the http://mylyn.org domain. Maybe that helps.

If that is the same instances that the unit tests are using (e.g.: https://mylyn.org/hudson-3.3.3, http://mylyn.org/jenkins-2.32.3/) it doesn't help.

Neither of them accept the ''crumbIssuer/api/json" request

**Comment by BeckerFrank** [comment]
Maybe the versions are to old to support that or we have never configure this.

**Comment by gnl42** [comment]
@BeckerFrank 
> Maybe the versions are to old to support that or we have never configure this.

Looks like 2.32.3 was released back in 2016 so I would say it's too old.
It also hasn't been configured to use the crumb security (CSRF not enabled)

**Comment by wimjongman** [comment]
> I would suggest merging the code and deal with any issues that crop up with a new ticket

Hey George. Cool! Do you want to take a look at the review comments before merging?

File: org.eclipse.mylyn-target/mylyn-e4.26.target
Comment (ruspl-afed) [reply org.eclipse.mylyn-target/mylyn-e4.26.target:129]: OMG. Let's postpone this improvement, I'd rather have a closer look on our .target later.

**Review by ruspl-afed — APPROVED** [review]
I've removed my comments regarding issues in .target, I need to have closer look later.

**Comment by gnl42** [comment]
> > I would suggest merging the code and deal with any issues that crop up with a new ticket
> 
> Hey George. Cool! Do you want to take a look at the review comments before merging?

I think I've addressed the comments to the best of my ability. When I tried to address the comments about the .target I ran into  unresolved references that I have no idea how to fix

**Comment by gnl42** [comment]
Don't merge yet. Started looking into the session Id problem and found some very not nice behaviour

**Comment by gnl42** [comment]
Solved the "sessionId" issue.

user/pw can run build now. Caveat is that if Jenkins is using StrictCrumbIssuer with a timeout set for the crumb one will be unable to run builds once the crumb expires. I can't see any way to have a new crumb created in this situation with how the build framework is designed.
Does not happen with a token, the recommended way, so I'm ok with this behaviour.

**Comment by wimjongman** [comment]
George, is the token used in the password field?

**Comment by gnl42** [comment]
Hi Wim,
   yes.

It's getting complicated. Depending on the system, authentication can be by:

  * Token
  * username/pw
  * username/token

George

**Comment by wimjongman** [comment]
Yes, it is. To be clear by 'system' do you mean systems like Jenkins or different Jenkins versions? Is the UI still using "Hudson"?

**Review by wimjongman — APPROVED** [review]

**Comment by gnl42** [comment]
Yes,

  * Token: Bamboo 8, GitHub, Jira 8
  * username/pw: Bamboo 8, Jenkins, Jira 8
  * username/token: Jira 10, Jenkins (modern)

The connector is advertised as 'Hudson (supports Jenkins)' but I don't 
see anything in the UI itself

**Comment by BeckerFrank** [comment]
> @BeckerFrank
> 
> > Maybe the versions are to old to support that or we have never configure this.
> 
> Looks like 2.32.3 was released back in 2016 so I would say it's too old. It also hasn't been configured to use the crumb security (CSRF not enabled)
Yes we should use a newer Jenkins version in our new test infrastructure (see Multipass setup supports >= 2_303_3)

**Comment by gnl42** [comment]
Could someone merge the PR?

**Pull request merged**
